### PR TITLE
ts: Add IDL seed types

### PR DESCRIPTION
**Pull request by Shiva953**
🤖 This pull request was created by https://app.gib.work/r/Ql3a0gKk/anchor 
 ⚡️ Payment for bounty is immediately sent to the contributor after merge.

### CODE REVIEW

File: ts/packages/anchor/src/idl.ts [hunk 1]
```diff
@@ -67,7 +67,16 @@ export type IdlPda = {
...
+// Based on the rust type defined in lang/sync/src/idl/types.rs and the definitions in lang/syn/src/idl/pda.rs
+export type IdlSeed =
+  | { kind: 'Const'; value: IdlConstant }
+  | { kind: 'Arg'; value: IdlSeedArg }
+  | { kind: 'Account'; value: IdlAccount };
+
+export interface IdlSeedArg {
+  type: IdlType;
+  path: string;
+}
```
Comment (acheroncrypto) [inline ts/packages/anchor/src/idl.ts:79]: Hey, thanks for your contribution but I don't think these types are correct based on the Rust definitions you've linked as a comment.

Comment (Shiva953) [reply ts/packages/anchor/src/idl.ts]: Thanks for the feedback. Updated the types, defined separate ones for `IdlSeedConst`, `IdlSeedArg` and `IdlSeedAccount` based on the rust definitions. Let me know the changes required.

File: ts/packages/anchor/src/idl.ts [hunk 2]
```diff
@@ -67,7 +67,38 @@ export type IdlPda = {
...
+  value: SerdeJsonValue;
+};
+
+export type SerdeJsonValue =
+  | null
+  | boolean
+  | number
+  | string
+  | SerdeJsonValue[]
+  | { [key: string]: SerdeJsonValue };
```
Comment (acheroncrypto) [inline ts/packages/anchor/src/idl.ts:88]: Thanks for adding this but I think we can just use `any` here and remove this type.

File: ts/packages/anchor/src/idl.ts [hunk 3]
```diff
@@ -67,7 +67,38 @@ export type IdlPda = {
...
+export type IdlSeedArg = {
+  kind: "arg";
+  type: IdlType;
+  path: string;
+};
+
+export type IdlSeedAccount = {
+  kind: "account";
+  type: IdlType;
+  account?: string; // Adjust the type based on the actual type of account in your use case
```
Comment (acheroncrypto) [inline ts/packages/anchor/src/idl.ts:99]:
```suggestion
  account?: string;
```

File: ts/packages/anchor/src/idl.ts [hunk 4]
```diff
@@ -67,7 +67,38 @@ export type IdlPda = {
   programId?: IdlSeed;
 };
 
-export type IdlSeed = any; // TODO
+// Based on the rust type defined in lang/sync/src/idl/types.rs and the definitions in lang/syn/src/idl/pda.rs
```
Comment (acheroncrypto) [inline ts/packages/anchor/src/idl.ts:70]:
```suggestion
```

**Review by acheroncrypto — COMMENTED** [review]
Thanks, looks great!

Before we merge, could you also add a [CHANGELOG](https://github.com/coral-xyz/anchor/blob/master/CHANGELOG.md#features) entry for this?

File: ts/packages/anchor/src/idl.ts (hunk 2 above)
Comment (Shiva953) [reply ts/packages/anchor/src/idl.ts:88]: sure!

**Comment by Shiva953** [comment]
@acheroncrypto any updates?

**Comment by acheroncrypto** [comment]
TS package has build errors: https://github.com/coral-xyz/anchor/actions/runs/7368700143/job/20074262110?pr=2752

**Comment by Shiva953** [comment]
The tests now pass, added the necessary missing types in `IdlSeedConst` and `IdlSeedArg`, also modified the error message a bit(since `accountDesc.pda.programId.kind` included inside the `throw New Error()` was resulting in type `never` for `accountDesc.pda.programId`)

Screenshot 2024-01-02 at 7 16 34 PM

**Review by acheroncrypto — CHANGES_REQUESTED** [review]
There is also a [formatting error](https://github.com/coral-xyz/anchor/actions/runs/7386584794/job/20106102432?pr=2752), you can fix it by running `yarn lint:fix`.

**Review by acheroncrypto — APPROVED** [review]
Thank you for the contribution!

**Pull request merged**